WordPress/gutenberg
# GitHub Actions: Create Release Draft when tagging version

# Description
Second stab at writing a workflow (GitHub action) to build Gutenberg upon creating a tag that starts with `v`, and creating a draft release with the build attached to it. Supersedes #19626.

The difference to #19626 is that:
- the release creation is now done in a separate job, and the plugin zip is retained as an artifact, and
- the release notes are extracted from the Changelog.

The (locally run) release script works pretty much as before -- only plugin zip creation and GitHub release creation have been removed (since they're now done by the new workflow).

Note that we're currently creating a release _draft_, meaning an extra step is required to publish the release: Clicking the 'Publish' button. While I find that extra safety net convenient, it is mainly for testing purposes. I'm happy to change it to publish the release right away -- it's only a [one-line change](https://github.com/WordPress/gutenberg/pull/27488/files#diff-40b47d855e2cffe931290bba37c1b8716642627cbe05ac859728cb95c5da7536R80) (plus a few instructions in the release scripts).

The new GH workflow uses the following GH actions:
- https://github.com/actions/create-release
- https://github.com/actions/download-artifact
- https://github.com/actions/upload-release-asset.

The main benefit of this PR is that the plugin build is decoupled from people's local systems and setups, yielding reproducible builds, and cutting down time spent running the release process locally (freeing up people's systems earlier). (One micro-optimization is e.g. that it's no longer necessary to enter a GH access token during the release process.)

## How has this been tested?

See here for an example release draft created by this action: https://github.com/ockham/gutenberg/releases/tag/untagged-f7f779e5825df118d22c

Corresponding workflow run: https://github.com/ockham/gutenberg/runs/1670519325?check_suite_focus=true

Or use the following instructions to test for yourself:

- Make sure you have your own fork of the Gutenberg repo, and that you've added it locally as a remote (E.g. `git remote add ockham git@github.com:ockham/gutenberg.git` if your GH username is `ockham` :slightly_smiling_face: ).
- Check out this PR's branch, and push the contents to your fork's `master` branch (you can reset later):
```
git push ockham add/gh-action-create-release:master
```
- Note the latest GB plugin release version (9.7.0 at the time of this writing).
- Delete the corresponding tag both locally (e.g. `git tag -d v9.7.0`) and in your fork `git push ockham :v9.7.0`.
- Recreate the tag, e.g. `git tag v9.7.0`.
- Push that tag to your fork: `git push --tags ockham`.
- Go to the 'Actions' tab of your fork's GH page (e.g. https://github.com/ockham/gutenberg/actions).
- Locate the action whose description reads "Build Gutenberg and upload to Draft Release", and click on it.
- It should take less than 5 minutes to build the plugin zip. You can watch that page as the action progresses.
- Eventually, a release draft will be created, and the plugin zip will be attached. You'll find it under 'Releases' (e.g. https://github.com/ockham/gutenberg/releases)
- Don't forget to clean up tags after running these commands, so you won't accidentally push tags that you solely created for testing to the actual Gutenberg repo later. See e.g. https://stackoverflow.com/questions/1841341/remove-local-git-tags-that-are-no-longer-on-the-remote-repository

(I've used my Gutenberg fork as an example, you can see the relevant action and release drafts from previous runs there.)

Alternatively, you can also use the local release script to create the tag and changelog. Make sure to modify `bin/plugin/config.js` to point to your forked repo!

## Types of changes
Build tools automation
